### PR TITLE
Cosmo as a dataclass

### PR DESCRIPTION
**Pull request by nstarman**
This PR changes `Cosmology` and its subclasses to `dataclasses.dataclass` objects for versions py3.10+.

**Why change?**

Well, this has long been the intended plan. `Cosmology` is structured very similarly to the dataclass paradigm (see `cosmology.Parameter`) but small differences meant that they were not compatible.

Dataclasses bring a number of advantages over normal classes, particularly for classes that are defined by their attributes (so not `dict`-like things).

1. straightforward declaration of the fields / attributes
2. no need to define most dunder methods.
3. built in static typing compatibility
4. support for match statements
5. easy to make immutable structures
6. introspection tools like `fields` and `asdict`, etc.
7. easier to dynamically define subclasses
8. deep support in transpilation / compilation libraries, e.g. mypyc.
9. ...

- [x] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/cosmology/flrw/tests/test_base.py
Comment (eerovaher) [inline astropy/cosmology/flrw/tests/test_base.py:350]: Tests shouldn't be using `np.all`, they should be using [`np.testing.assert_array_equal()`](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_array_equal.html).

File: astropy/cosmology/flrw/tests/test_w0cdm.py
Comment (eerovaher) [inline astropy/cosmology/flrw/tests/test_w0cdm.py]: What is causing the order of some of the parameters to change and could it be avoided?

File: astropy/cosmology/_utils.py
Comment (eerovaher) [inline astropy/cosmology/_utils.py:95]: But any `type` is also an `object`.

File: astropy/cosmology/tests/test_utils.py
Comment (eerovaher) [inline astropy/cosmology/tests/test_utils.py:3]: Why?

File: astropy/cosmology/utils.py
Comment (eerovaher) [inline astropy/cosmology/utils.py:3]: Why?

File: astropy/cosmology/flrw/tests/test_w0cdm.py
Comment (eerovaher) [inline astropy/cosmology/flrw/tests/test_w0cdm.py]: Why did you bother introducing `expected` for a couple of commits? This just clutters the diff.

File: astropy/cosmology/_utils.py
Comment (eerovaher) [inline astropy/cosmology/_utils.py:104]: Is there an import loop?

File: astropy/cosmology/core.py
Comment (eerovaher) [inline astropy/cosmology/core.py:82]: Does this not fit on one line?
```suggestion
        return self.default if instance is None else instance._name
```

File: astropy/cosmology/flrw/base.py
Comment (eerovaher) [inline astropy/cosmology/flrw/base.py:1586]:
```suggestion
        if not to_nonflat and kwargs.get("Ode0", None) is not None:
```

**Review by eerovaher — COMMENTED** [review]
I'm not convinced that this should be included in 6.0. One of the main advantages of converting `Cosmology` to a `dataclass` would be that many dunder methods would automatically generated, but right now `Cosmology` has to maintain compatibility with Python 3.9, so the dunder methods still have to be implemented (although now in a `if PYTHON_LT_3_10:` block). Furthermore, having the base classes of `Cosmology` depend on the Python version does not sound sensible to me.

File: astropy/cosmology/parameter/tests/test_parameter.py
Comment (nstarman) [reply astropy/cosmology/parameter/tests/test_parameter.py:212]: Arguably this should have been removed long ago. I added this test when I was first learning implementing the descriptors on Cosmology. But the behaviour of descriptors is set by the language and especially enforced in dataclasses which share Field objects. There's no need to test this.

File: astropy/cosmology/parameter/tests/test_parameter.py
Comment (nstarman) [reply astropy/cosmology/parameter/tests/test_parameter.py:444]: It only gets tested once now, not repeated for each cosmology (they use the test mixin from which it was moved).

Comment (nstarman) [reply astropy/cosmology/parameter/tests/test_parameter.py:444]: I guess I'm saying yes 😆

File: astropy/cosmology/core.py
Comment (nstarman) [reply astropy/cosmology/core.py]: But ``MetaData`` will change the type to match what's passed.
However `Cosmology` used to do ``meta.update`` which meant it stayed an ``OrderedDict``. So it's necessary to cast to an ``OrderedDict`` now.

File: astropy/cosmology/core.py
Comment (nstarman) [reply astropy/cosmology/core.py]: I don't see this in ``core``. Were you looking at a specific commit ?

File: astropy/cosmology/core.py
Comment (nstarman) [reply astropy/cosmology/core.py:197]: "`Cosmology.__post_init__` is an empty method in an abstract base class, but has no abstract decorator Ruff[B027](https://beta.ruff.rs/docs/rules/empty-method-without-abstract-decorator)"

Comment (nstarman) [reply astropy/cosmology/core.py:197]: ``super(). __post_init__ `` errors if there's not ``__post_init__`` method so it's necessary to have the empty method. And thus the noqa.

File: astropy/cosmology/parameter/tests/test_parameter.py
Comment (eerovaher) [reply astropy/cosmology/parameter/tests/test_parameter.py:444]: That sounds reasonable, but it's not obvious from the diff and it's not explained in the commit message, so how am I supposed to know that?

File: astropy/cosmology/core.py
Comment (eerovaher) [reply astropy/cosmology/core.py]: I was indeed looking at the commits one-by-one, but the question remains.

File: astropy/cosmology/core.py
Comment (eerovaher) [reply astropy/cosmology/core.py:197]: But why are the subclasses calling `super().__post_init__()` if it doesn't do anything?

File: astropy/cosmology/core.py
Comment (nstarman) [reply astropy/cosmology/core.py]: There's no ``__post_init__`` to inherit. It's not on `object`

File: astropy/cosmology/core.py
Comment (nstarman) [reply astropy/cosmology/core.py:197]: Because they get mixed in with `FlatFLRWMixin` to make the flat cosmologies. Both have ``__post_init__`` so it's necessary to ensure stuff passes up the chain.

File: astropy/cosmology/flrw/tests/test_w0cdm.py
Comment (nstarman) [reply astropy/cosmology/flrw/tests/test_w0cdm.py]: Unfortunately not. Dataclasses strictly order the fields. And this is reflected in the string representation. The disadvantage is that it can change the representation, like here, and cannot be re-ordered. The advantage is that it is entirely deterministic and predictable. The previous repr PR brought `Cosmology` into close agreement with how dataclass repr looks, but I don't reorder the parameters, saving that for this PR.

File: astropy/cosmology/_utils.py
Comment (nstarman) [reply astropy/cosmology/_utils.py:104]: Yeah, there was. Utils should not depend on anything else.

File: astropy/cosmology/parameter/tests/test_parameter.py
Comment (nstarman) [reply astropy/cosmology/parameter/tests/test_parameter.py:212]: Sure.

File: astropy/cosmology/parameter/tests/test_parameter.py
Comment (nstarman) [reply astropy/cosmology/parameter/tests/test_parameter.py:444]: I did this diff as part of #15549.

File: astropy/cosmology/parameter/tests/test_parameter.py
Comment (nstarman) [reply astropy/cosmology/parameter/tests/test_parameter.py:212]: #15549

File: astropy/cosmology/_utils.py
Comment (nstarman) [reply astropy/cosmology/_utils.py:95]: Yes, but it's still an informative message for the user, which in this case is us maintainers for this private function. 
`obj` can be an instance of a class or the class itself.

File: astropy/cosmology/core.py
Comment (nstarman) [reply astropy/cosmology/core.py:82]: I added comments instead explaining the different access methods.

File: astropy/cosmology/flrw/tests/test_w0cdm.py
Comment (nstarman) [reply astropy/cosmology/flrw/tests/test_w0cdm.py]: Removed in the git history.

File: astropy/cosmology/utils.py
Comment (nstarman) [reply astropy/cosmology/utils.py]: Removed from the git history. Legacy of a squash where a util had type hints.

File: astropy/cosmology/core.py
Comment (nstarman) [reply astropy/cosmology/core.py]: They are only necessary from a static typing perspective given that mypy has little understanding of mixin classes. So I'll remove it here.

File: astropy/cosmology/flrw/tests/test_base.py
Comment (nstarman) [reply astropy/cosmology/flrw/tests/test_base.py]: ✅

File: astropy/cosmology/tests/test_utils.py
Comment (nstarman) [reply astropy/cosmology/tests/test_utils.py]: Removed.

**Comment by nstarman** [comment]
v6.1 is fine I guess. 

https://numpy.org/neps/nep-0029-deprecation_policy.html says "On Apr 05, 2024 drop support for Python 3.9 (initially released on Oct 05, 2020)". Astropy v6.1 won't be released until after April, right? So we can bump our python version  for v6.1, making this PR significantly simpler.

**Comment by pllim** [comment]
Yes, "major/minor" release is still every 6 months or so. See https://github.com/astropy/astropy-APEs/blob/main/APE2.rst#detailed-description

**Comment by eerovaher** [comment]
It would be best to put this on hold until support for Python 3.9 is dropped.

**Comment by nstarman** [comment]
Ok. Then we can bump Python for v6.1 and get this in.

**Comment by saimn** [comment]
https://github.com/astropy/astropy/wiki/Release-Calendar needs some updates but yes, 6.1 should happen early May 2024.

**Comment by pllim** [comment]
BTW, if you want to bump Python, please do that in a separate PR, not here. Thanks!

File: astropy/cosmology/core.py
Comment (neutrinoceros) [inline astropy/cosmology/core.py:141]: maybe I'm only asking because I don't have a clear idea of how this class already works but this looks like a breaking change, as well as a Python version-dependent behaviour (but that's secondary since 3.9 is going away soon anyway).
I'm worried that the breaking change would have been fine in 6.0 but not for 6.1.

**Review by neutrinoceros — COMMENTED** [review]
My understanding is that this PR will become a whole lot easier to review later so I haven't gone over all of it yet, but I do have an important comment about what looks like a breaking change in the public API (namely how one is allowed to instantiate `Cosmology`).

Furthermore, and out of curiosity, can I ask for a short explanation of what makes Python 3.10's data classes better for *existing* code ? I've glossed over the [release notes](https://docs.python.org/3/whatsnew/3.10.html#dataclasses) and both improvements from this version look (to me ?) like breaking if applied after the fact (namely support for `__slots__` and keyword-only fields.

File: astropy/cosmology/core.py
Comment (nstarman) [reply astropy/cosmology/core.py:141]: You're technically correct that this PR changes `name` and `meta` from positional-or-keyword arguments to keyword-only arguments, which is a change. However in practice this is not a change. I couldn't find a single example in Astropy, on GitHub, nor GitLab where anyone ever passes these arguments positionally (the closest is using `inspect.BoundArguments()` which auto-splits args and kwargs, but this is not affected by the change for that same reason). There's a good reason no one positionally passes `name` and `meta`: in all the non-abstract classes `name` and `meta` appear at the end of the `__init__` signature. That, and since different classes have different number of preceding arguments, in practice these have always been keyword arguments. As such I think it's fine for v6.1.

Comment (nstarman) [reply astropy/cosmology/core.py:141]: As for the Python-version dependent behavior, this is actually why this PR has to wait for Py3.10. Earlier forms of dataclasses don't understand how to rearrange non-defaulted parameters based on redefinitions when inheriting from a class with those same parameters. Ie you can't redefine `name` and `meta` in a subclass to put them at the end if you have a parameter `H0` that doesn't have a default value.

**Comment by nstarman** [comment]
Rebased on #15706 and #15603. when those are merged I'll re-mark this PR as ready for review.

**Comment by nstarman** [comment]
> Furthermore, and out of curiosity, can I ask for a short explanation of what makes Python 3.10's data classes better for _existing_ code ?

`dataclasses.dataclass` objects have a number of distinct advantages

1. Less code => reduced code debt
2. Natural definition of immutability
3. Better introspection (annotations and tools from `dataclasses`)
4. Better I/O (tools from `dataclasses`)
5. Enforced good code structure.
6. ensure compatibility with `mypyc` transpilation.

**Comment by neutrinoceros** [comment]
> why would we not make a class a dataclass?

A concern I heard about data classes is startup time overhead.
There are also [alternatives like `attrs`](https://www.attrs.org/en/stable/why.html#data-classes), which is historically a precursor to data classes, but it is still actively developed as a third party package.
But I don't think we need to worry about the startup overhead too much for a library, and I just mentioned attrs for good measure, but if we can live happily with the standard lib implementation, no need to reach out for another third party dependency !

**Comment by nstarman** [comment]
@eerovaher @WilliamJamieson @neutrinoceros, now that #15603 is in, this PR is ready as well! It will be great to finally upgrade the architecture of `Cosmology`. This should make a lot of other features easier to implement.

**Review by neutrinoceros — COMMENTED** [review]
I'll set aside some time to finish this review tomorrow, in the mean time here are a couple (very) small questions.

File: astropy/cosmology/tests/test_core.py
Comment (neutrinoceros) [inline astropy/cosmology/tests/test_core.py:311]: In contrast to a previous comment, I agree this should be dropped.

File: astropy/utils/compat/misc.py
Comment (neutrinoceros) [inline astropy/utils/compat/misc.py:14]: also a trivial conflict here

File: docs/changes/cosmology/15484.api.rst
Comment (neutrinoceros) [inline docs/changes/cosmology/15484.api.rst:2]: I think linking https://docs.python.org/3/library/dataclasses.html should be safe here ?

**Review by neutrinoceros — COMMENTED** [review]
This time I was able to do a complete read through. This looks outstanding !

A broad (but very minor) comment is that I'm not sure about the name `dataclass_decorator`. How about maybe something more descriptive of its application domain like `cosmology_dataclass` ?

File: astropy/cosmology/flrw/tests/test_base.py
Comment (nstarman) [reply astropy/cosmology/flrw/tests/test_base.py:1132]: So mypy was complaining on my IDE that the subclass violates Liskov.

File: docs/changes/cosmology/15484.api.rst
Comment (nstarman) [reply docs/changes/cosmology/15484.api.rst:2]: @pllim, thoughts? I know we try to minimize links in our changelog

File: astropy/utils/compat/misc.py
Comment (nstarman) [reply astropy/utils/compat/misc.py]: Rebased and fixed, thanks.

File: astropy/cosmology/tests/test_core.py
Comment (nstarman) [reply astropy/cosmology/tests/test_core.py]: Rebased and fixed.

File: astropy/cosmology/parameter/tests/test_parameter.py
Comment (nstarman) [reply astropy/cosmology/parameter/tests/test_parameter.py:173]: No. There's just little reason to test this since it's in core Python. We don't override ``__setattr__`` anywhere, so...

File: astropy/cosmology/parameter/tests/test_descriptors.py
Comment (nstarman) [reply astropy/cosmology/parameter/tests/test_descriptors.py:140]: Can do.

Comment (nstarman) [reply astropy/cosmology/parameter/tests/test_descriptors.py:140]: Done.

File: astropy/cosmology/flrw/tests/test_base.py
Comment (neutrinoceros) [reply astropy/cosmology/flrw/tests/test_base.py:1091]: ok, seems reasonable to me

File: astropy/cosmology/flrw/tests/test_base.py
Comment (neutrinoceros) [reply astropy/cosmology/flrw/tests/test_base.py:1132]: Makes sense, thanks !

File: astropy/cosmology/core.py
Comment (nstarman) [reply astropy/cosmology/core.py:61]: Done.

File: docs/whatsnew/6.1.rst
Comment (nstarman) [reply docs/whatsnew/6.1.rst]: Removed.

File: astropy/cosmology/parameter/tests/test_parameter.py
Comment (nstarman) [reply astropy/cosmology/parameter/tests/test_parameter.py:19]: Which line?

File: astropy/cosmology/parameter/tests/test_parameter.py
Comment (neutrinoceros) [reply astropy/cosmology/parameter/tests/test_parameter.py:173]: that's fine !

**Review by neutrinoceros — APPROVED** [review]
With your latest batch of replies, I'm happy to sign this off !

**Comment by nstarman** [comment]
@eerovaher @WilliamJamieson, the feature freeze is coming up. What do y'all think of this PR?

**Comment by eerovaher** [comment]
I agree we should try to get this merged before the feature freeze. I'll make time to have a look at it tomorrow. Luckily @neutrinoceros has already reviewed it, so I don't expect that any big changes are going to be needed.

File: docs/changes/cosmology/15484.api.rst
Comment (eerovaher) [reply docs/changes/cosmology/15484.api.rst:2]: I'd say that because there is a What's New entry then the change log entry can be shorter.

File: docs/whatsnew/6.1.rst
Comment (eerovaher) [inline docs/whatsnew/6.1.rst:49]: This could have been written with less effort, but of course the effort has already been spent.
```suggestion
|Cosmology| as a :func:`~dataclasses.dataclass`
```

File: docs/whatsnew/6.1.rst
Comment (eerovaher) [inline docs/whatsnew/6.1.rst:56]: Why are the doctests being skipped?

File: docs/whatsnew/6.1.rst
Comment (eerovaher) [inline docs/whatsnew/6.1.rst:78]: Are there too many fields to write out all of them? If some fields should be skipped then I'd rather keep the fields in the beginning and end and only skip the ones in the middle.

**Review by eerovaher — COMMENTED** [review]
I do have a few comments, but it should be possible to address them before the feature freeze.

Do you plan editing the commit history yourself? Currently there are two different commits that both `Make Cosmology a dataclass`, so a normal merge doesn't seem to be the right thing to do, but you don't want this to be automatically squashed either.

**Comment by nstarman** [comment]
> Do you plan editing the commit history yourself? Currently there are two different commits that both Make Cosmology a dataclass, so a normal merge doesn't seem to be the right thing to do, but you don't want this to be automatically squashed either.

Yes, this is a product of an interactive rebase. I'll fix it up.

File: astropy/cosmology/flrw/base.py
Comment (nstarman) [reply astropy/cosmology/flrw/base.py:171]: The annotation is needed to make this a dataclass field. It's interesting, but `dataclass` disambiguates between descriptors that are vs aren't dataclass fields by the present of the annotation.

File: astropy/cosmology/core.py
Comment (nstarman) [reply astropy/cosmology/core.py]: ✅

File: astropy/cosmology/core.py
Comment (nstarman) [reply astropy/cosmology/core.py]: ✅

File: docs/whatsnew/6.1.rst
Comment (nstarman) [reply docs/whatsnew/6.1.rst:78]: It was just to emphasize `'newfield'`.

Comment (nstarman) [reply docs/whatsnew/6.1.rst:78]: I added the other fields.

File: docs/whatsnew/6.1.rst
Comment (nstarman) [reply docs/whatsnew/6.1.rst]: ✅

File: astropy/cosmology/_io/model.py
Comment (nstarman) [reply astropy/cosmology/_io/model.py:134]: It's just out of scope. This function has to comply with requirements for making Astropy Models.

File: astropy/cosmology/_utils.py
Comment (nstarman) [reply astropy/cosmology/_utils.py:102]: ✅

File: astropy/cosmology/parameter/tests/test_parameter.py
Comment (nstarman) [reply astropy/cosmology/parameter/tests/test_parameter.py:19]: Oh. A cleanup PR would be nice to do soon!

File: docs/changes/cosmology/15484.api.rst
Comment (nstarman) [reply docs/changes/cosmology/15484.api.rst]: Done!

**Review by eerovaher — APPROVED** [review]

**Comment by nstarman** [comment]
image

**Pull request merged**
